Hosted launch: Worker queues and executor updates (#2307)
astronomer/docs: astro/configure-worker-queues.md

import TabItem from '@theme/TabItem';
import {siteVariables} from '@site/src/versions';

> By default, all tasks using the Celery executor run in a `default` worker queue that does not require configuration or code. You can create additional worker queues to enable multiple worker types or configurations for different groups of tasks, and assign tasks to queues in your DAG code. For more information about Airflow executors on Astro, see [Manage executors](executors.md).

Reviewer: @jwitz not sure what we mean by this - `does not require configuration or code` . Default worker queue can also be configured based on the default workloads a user wants to run, like max tasks , max workers etc., which is same for any other worker Queue. And I think by `code`, we mean to say we do not require any changes to DAG code for using it. 

Given the above, can we rephrase something like "By default, all tasks using the Celery executor run in a `default` worker queue without any code change in your DAG. You can create additional worker queues to enable multiple worker types or configurations for different groups of tasks, and assign tasks to queues by changing your DAG code. See this [example](https://docs.astronomer.io/astro/configure-worker-queues?tab=classicoperator#assign-tasks-to-a-worker-queue) for details.

By configuring multiple worker queues and assigning tasks to these queues based on the requirements of the tasks, you can enhance the performance, reliability, and throughput of your Deployment. For example, consider the following scenario:

- You are running Task A and Task B in a Deployment.
- Task A and Task B are dependent on each other, so they need to run in the same Deployment.
- Task A is a long-running task that uses a lot of CPU and memory, while Task B is a short-running task that uses minimal amounts of CPU and memory.

> You can assign Task A to a worker queue that is configured to use the A20 worker type that's optimized for running larger tasks. Then, you can assign Task B to a worker queue that is configured to use the A5 worker type that's smaller and optimized for general usage.

Reviewer: ```suggestion
You can assign Task A to a worker queue that is configured to use the A20 worker type that's optimized for running compute-heavy tasks. Then, you can assign Task B to a worker queue that is configured to use the A5 worker type that's smaller and optimized for general usage.
```

## Worker queue settings

You can configure each worker queue on Astro with the following settings:

- **Name:** The name of your worker queue. Use this name to assign tasks to the worker queue in your DAG code. Worker queue names must consist only of lowercase letters and hyphens. For example, `machine-learning-tasks` or `short-running-tasks` or `high-cpu`.

> - **Worker Type:** The size of workers in the worker queue, for example A5 or A20. A worker’s total available CPU, memory, storage, and GPU is defined by its worker size. 

Reviewer 2: We don't support GPUs yet so I would remove that part altogether. 

Also, for now, all 3 worker types have 5 GiB ephemeral storage. 

So maybe this should say:

`A worker’s total available CPU and memory is defined by its worker size. Each Worker Type comes with 5 GiB of ephemeral storage.`

Author: I'm making a note to link to the Hosted resource reference on worker info for this detail.

> - **Max Tasks per Worker:** The maximum number of tasks that a single worker can run at a time. If the number of queued and running tasks exceeds this number, a new worker is added to run the remaining tasks. This value is equivalent to [worker concurrency](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#worker-concurrency) in Apache Airflow. It is 16 by default.

Reviewer 2: This part is different in Hosted. The setting is labeled as CONCURRENCY ( instead of Max Tasks per Worker ). In addition, the default is no longer 16, we are changing the default based on the size of the Worker Type ( try this out in dev and you'll see ).

If you choose A5 the default will change to 5
A10 --> 10
A20 --> 20

This is an advanced configuration and we don't recommend a user changing it unless they are sure.

> - **Worker Count**: The minimum and maximum number of workers that can run at a time.  The number of running workers changes regularly based on Maximum Tasks per Worker and the current number of tasks in a `queued` or `running` state. By default, the minimum number of workers is 1 and the maximum is 10.

Reviewer 2: This is labeled as MIN # WORKERS / MAX # WORKERS in Hosted.

In Hosted the min is also 0, so that's different too ( It also defaults to 0 ).

> - **Min # Workers / Max # Workers**: The minimum and maximum number of workers that can run at a time.  The number of running workers changes regularly based on **Concurrency** and the current number of tasks in a `queued` or `running` state. 

Reviewer: Feel like we should mention here about auto-scaling - "Your workers will auto-scale up to the max number of workers based on the concurrency and number of tasks in `queued` or `running` state."

:::info Alternative Astro Hybrid setup 

> On Astro Hybrid clusters, worker type is defined as a node instance type that is supported by the cloud provider of your cluster. For example, a worker type might be `m5.2xlarge` or `c6i.4xlarge` for a Deployment running on a Hybrid AWS cluster hosted on your cloud. Actual worker size is equivalent to the total capacity of the worker type minus Astro’s system overhead.

Reviewer: would be helpful to link here to the Cluster Setting Reference which gives the details about the amount of cpu and memory available minus the system overhead.

> Your Organization can enable up to 10 different worker types for each Hybrid cluster. After a worker type is enabled on an Astro Hybrid cluster, the worker type becomes available to any Deployment in that cluster and appears in the **Worker Type** menu of the Cloud UI.

Reviewer: this is not true for Azure, it only have 6 worker types - https://docs.astronomer.io/astro/resource-reference-azure#supported-worker-node-pool-instance-types

Author: Yes Azure only has 6 worker types, but if we had more you could technically enable up to 10. This is more of a message on cluster limits than it is about how many worker types we support in general.

Reviewer: @jwitz make sense. So then let's rephrase to "up to 10 additional different worker types". Because total 11 are supported in GCS and AWS.

1. Review the list of supported worker types for your cloud provider. See [AWS](resource-reference-aws.md#worker-node-types), [Azure](resource-reference-azure.md#worker-node-size-resource-reference), or [GCP](resource-reference-gcp.md#worker-node-size-resource-reference).

> 2. Contact [Astronomer support](https://cloud.astronomer.io/support) with the name of the worker type(s) you want to enable for your cluster. For example, `m6i.2xlarge`.

Reviewer: ```suggestion
2. Contact [Astronomer support](https://cloud.astronomer.io/support) with the name of the worker type(s) you want to enable for your cluster along with the cluster name. For example, `m6i.2xlarge` for my-AWS-dev-Cluster.
```

For more information on requesting cluster changes, see [LINK].

:::

## Default worker queue

> Each Deployment requires a worker queue named `default` to run tasks. Tasks that are not assigned to a worker queue in your DAG code are executed by workers in the default worker queue.

Reviewer: ```suggestion
Each Deployment requires a worker queue named `default` to run tasks. Tasks that are not specifically assigned to a worker queue in your DAG code are executed by workers in the default worker queue.
```

Author: This word seems unnecessary. Is there any way to assign a task to a worker queue thats unspecifically?

Reviewer: it is not, but we have to specify a "queue" to assign it to a particular worker Queue, else it just goes to default. So wanted to emphasize , if you change your DAG code and add a `queue` parameter, then only tasks are assigned to a different queue. Most of the OSS users are unfamiliar with Queue or Worker Queue concept, so I think it is important to make it clear.
